<robot
  name="mobile_robot">
  <gazebo reference="chassis">
    <material>Gazebo/Red</material>
  </gazebo>
  <gazebo reference="right_wheel">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="left_wheel">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="front_wheel">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="holder">
    <material>Gazebo/Red</material>
  </gazebo>
  <gazebo reference="left_sensor">
    <material>Gazebo/White</material>
  </gazebo>
  <gazebo reference="right_sensor">
    <material>Gazebo/white</material>
  </gazebo>
  <gazebo reference="front_sensor">
    <material>Gazebo/White</material>
  </gazebo>
  <link
    name="chassis">
    <inertial>
      <origin
        xyz="-0.147460714742137 -0.0302705388894093 0.0197397389092321"
        rpy="0 0 0" />
      <mass
        value="5.28470193836214" />
      <inertia
        ixx="0.0172196936151484"
        ixy="0.00207025336305734"
        ixz="-0.000349261147912065"
        iyy="0.0883683584210481"
        iyz="8.82258104464462E-06"
        izz="0.0760293825469887" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://mobile_robot_description/meshes/chassis.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.792156862745098 0.819607843137255 0.933333333333333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://mobile_robot_description/meshes/chassis.STL" />
      </geometry>
    </collision>
  </link>
  <link
    name="right_wheel">
    <inertial>
      <origin
        xyz="-0.0216872360293876 4.91273688396632E-15 2.49212142519606E-07"
        rpy="0 0 0" />
      <mass
        value="0.501600220937669" />
      <inertia
        ixx="0.00143563367410199"
        ixy="-1.52145813469872E-16"
        ixz="-5.81437685180484E-10"
        iyy="0.000767786397856862"
        iyz="-1.0842021724855E-19"
        izz="0.000767786397888015" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://mobile_robot_description/meshes/right_wheel.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.792156862745098 0.819607843137255 0.933333333333333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://mobile_robot_description/meshes/right_wheel.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="right_wheel_joint"
    type="continuous">
    <origin
      xyz="-0.232516893895987 -0.00992377145207874 -0.0943355033260453"
      rpy="2.06081793355236 -1.5707963267949 0" />
    <parent
      link="chassis" />
    <child
      link="right_wheel" />
    <axis
      xyz="1 0 0" />
  </joint>
  <link
    name="left_wheel">
    <inertial>
      <origin
        xyz="0.0211606147695255 2.13717932240343E-15 2.49212143400845E-07"
        rpy="0 0 0" />
      <mass
        value="0.50160022093767" />
      <inertia
        ixx="0.00143563367410199"
        ixy="6.69005152138131E-17"
        ixz="-5.81437083891871E-10"
        iyy="0.000767786397856863"
        iyz="-1.35525271560688E-20"
        izz="0.000767786397888016" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://mobile_robot_description/meshes/left_wheel.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.792156862745098 0.819607843137255 0.933333333333333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://mobile_robot_description/meshes/left_wheel.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="left_wheel_joint"
    type="continuous">
    <origin
      xyz="-0.232516893895987 -0.0099237714521887 0.132816645875042"
      rpy="1.36214888437792 -1.5707963267949 0" />
    <parent
      link="chassis" />
    <child
      link="left_wheel" />
    <axis
      xyz="1 0 0" />
  </joint>
  <link
    name="holder">
    <inertial>
      <origin
        xyz="-7.91033905045424E-16 0.00167070427770837 0"
        rpy="0 0 0" />
      <mass
        value="0.0145605637508094" />
      <inertia
        ixx="7.06262048877384E-06"
        ixy="1.87686475973523E-18"
        ixz="-1.05879118412685E-21"
        iyy="3.15999848872344E-06"
        iyz="1.11672116017902E-21"
        izz="9.78305610352119E-06" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://mobile_robot_description/meshes/holder.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.792156862745098 0.819607843137255 0.933333333333333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://mobile_robot_description/meshes/holder.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="holder_joint"
    type="fixed">
    <origin
      xyz="0.0943608907023675 0 0.0192556511477779"
      rpy="-3.14159265358973 -1.44058362416008 1.98210909358335E-15" />
    <parent
      link="chassis" />
    <child
      link="holder" />
    <axis
      xyz="0 1 0" />
  </joint>
  <link
    name="front_wheel">
    <inertial>
      <origin
        xyz="-0.0005911395494715 -6.76188943588008E-08 -7.21490811124437E-06"
        rpy="0 0 0" />
      <mass
        value="0.07939886801277" />
      <inertia
        ixx="6.61465252204961E-05"
        ixy="6.02810543812606E-12"
        ixz="-2.05702274983285E-10"
        iyy="3.52805745190462E-05"
        iyz="5.7216274092266E-11"
        izz="3.52806223740297E-05" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://mobile_robot_description/meshes/front_wheel.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.792156862745098 0.819607843137255 0.933333333333333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://mobile_robot_description/meshes/front_wheel.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="front_wheel_joint"
    type="continuous">
    <origin
      xyz="0 -0.0248262285478683 0"
      rpy="-1.08610219202275 -2.49800180540668E-16 7.86938930123298E-17" />
    <parent
      link="holder" />
    <child
      link="front_wheel" />
    <axis
      xyz="1 0 0" />
  </joint>
    <gazebo>
      <plugin name="differential_drive_controller" filename="libgazebo_ros_diff_drive.so">
        <wheelSeparation>0.2271521492</wheelSeparation>
        <wheelDiameter>0.265622</wheelDiameter>
        <leftJoint>left_wheel_joint</leftJoint>
        <rightJoint>right_wheel_joint</rightJoint>
        <robotBaseFrame>chassis</robotBaseFrame>
        <publishWheelJointState>true</publishWheelJointState>
        <legacyMode>fasle</legacyMode>
      </plugin>
    </gazebo>


    <link name="left_sensor">
      <visual>
        <geometry>
          <cylinder length="0.01" radius="0.01"/>
        </geometry>
      </visual>
    </link>
    <joint name="left_sensor_joint" type="fixed">
      <parent link="chassis"/>
      <child link="left_sensor"/>
      <origin xyz="0 -0.07 0.07" rpy="1.57 -1.57 0"/>
    </joint>


<gazebo reference="left_sensor">
   <sensor type="ray" name="ir_left">
      <pose>0 0 0 0 0 0</pose>
      <update_rate>50</update_rate>
      <ray>
         <scan>
            <horizontal>
               <samples>1</samples>
               <resolution>1.0</resolution>
               <min_angle>-0.01</min_angle>
               <max_angle>0.01</max_angle>
            </horizontal>
            <vertical>
               <samples>1</samples>
               <resolution>1</resolution>
               <min_angle>-0.01</min_angle>
               <max_angle>0.01</max_angle>
            </vertical>
         </scan>
         <range>
            <min>0.01</min>
            <max>3.75</max>
            <resolution>0.02</resolution>
         </range>
      </ray>
      <plugin filename="libgazebo_ros_range.so" name="gazebo_ros_range">
         <gaussianNoise>0.005</gaussianNoise>
         <alwaysOn>true</alwaysOn>
         <updateRate>5</updateRate>
         <topicName>/ir_left</topicName>
         <frameName>left_sensor</frameName>
         <visualize>true</visualize>
         <radiation>infrared</radiation>
         <fov>0.02</fov>
      </plugin>
   </sensor>
 </gazebo>
   <link name="right_sensor">
     <visual>
       <geometry>
         <cylinder length="0.01" radius="0.01"/>
       </geometry>
     </visual>
   </link>
   <joint name="right_sensor_joint" type="fixed">
     <parent link="chassis"/>
     <child link="right_sensor"/>
     <origin xyz="0 -0.07 -0.03" rpy="1.57 1.57 0"/>
   </joint>
   <gazebo reference="right_sensor">
      <sensor type="ray" name="ir_right">
         <pose>0 0 0 0 0 0</pose>
         <update_rate>50</update_rate>
         <ray>
            <scan>
               <horizontal>
                  <samples>1</samples>
                  <resolution>1.0</resolution>
                  <min_angle>-0.01</min_angle>
                  <max_angle>0.01</max_angle>
               </horizontal>
               <vertical>
                  <samples>1</samples>
                  <resolution>1</resolution>
                  <min_angle>-0.01</min_angle>
                  <max_angle>0.01</max_angle>
               </vertical>
            </scan>
            <range>
               <min>0.01</min>
               <max>3.75</max>
               <resolution>0.02</resolution>
            </range>
         </ray>
         <plugin filename="libgazebo_ros_range.so" name="gazebo_ros_range">
            <gaussianNoise>0.005</gaussianNoise>
            <alwaysOn>true</alwaysOn>
            <updateRate>5</updateRate>
            <topicName>/ir_right</topicName>
            <frameName>right_sensor</frameName>
            <visualize>true</visualize>
            <radiation>infrared</radiation>
            <fov>0.02</fov>
         </plugin>
      </sensor>
</gazebo>



<link name="front_sensor">
  <visual>
    <geometry>
      <cylinder length="0.01" radius="0.01"/>
    </geometry>
  </visual>
</link>
<joint name="front_sensor_joint" type="fixed">
  <parent link="chassis"/>
  <child link="front_sensor"/>
  <origin xyz="0.07 -0.07 0.02" rpy="1.57 0 0"/>
</joint>
<gazebo reference="front_sensor">
   <sensor type="ray" name="ir_front">
      <pose>0 0 0 0 0 0</pose>
      <update_rate>50</update_rate>
      <ray>
         <scan>
            <horizontal>
               <samples>1</samples>
               <resolution>1.0</resolution>
               <min_angle>-0.01</min_angle>
               <max_angle>0.01</max_angle>
            </horizontal>
            <vertical>
               <samples>1</samples>
               <resolution>1</resolution>
               <min_angle>-0.01</min_angle>
               <max_angle>0.01</max_angle>
            </vertical>
         </scan>
         <range>
            <min>0.01</min>
            <max>3.75</max>
            <resolution>0.02</resolution>
         </range>
      </ray>
      <plugin filename="libgazebo_ros_range.so" name="gazebo_ros_range">
         <gaussianNoise>0.005</gaussianNoise>
         <alwaysOn>true</alwaysOn>
         <updateRate>5</updateRate>
         <topicName>/ir_front</topicName>
         <frameName>front_sensor</frameName>
         <visualize>true</visualize>
         <radiation>infrared</radiation>
         <fov>0.02</fov>
      </plugin>
   </sensor>
</gazebo>

<gazebo>
    <plugin name="imu_plugin" filename="libgazebo_ros_imu.so">
      <alwaysOn>true</alwaysOn>
      <bodyName>chassis</bodyName>
      <topicName>imu</topicName>
      <serviceName>imu_service</serviceName>
      <gaussianNoise>0.0</gaussianNoise>
      <updateRate>20.0</updateRate>
    </plugin>
  </gazebo>
</robot>
